66 micro loans part 1

ZacTredger opened the pull request:
I renamed the branch and expected the conversation from the last PR to be preserved, but GitHub strangely didn't understand that it was the same branch, so I had to create a new PR. I resolved all the previous comments though!

This PR also includes the *acceptance*(/'granting') of microloans!

Thanks all
Z

ZacTredger committed:
Add more global constants to configuration file

Some constants (like DEFAULT_CURRENCY) were hard-coded in multiple places.

ZacTredger committed:
Upgrade to Junit 5

I've migrated the JUnit 4 tests, but I just deleted AllTests, because I could
get `mvn test` to run the JUnit 5 tests without needing it. Please someone tell
me if this clobbers anything! I did a lot of research, but I still feel like I'm
not in familiar territory!

JUnit 5 will enable per-test parameterisation, which will be needed by the
microloan implementation.

ZacTredger committed:
Add regression test for DEFAULT when logged out

juliano60 commented:
@ZacTredger Shall have a read before the end of the day. Obviously quite a few files touched, but since we don't have the luxury of time...Note that it is out of synch with "main".

Regards,
Jules

ZacTredger commented:
> @ZacTredger Shall have a read before the end of the day. Obviously quite a few files touched, but since we don't have the luxury of time...Note that it is out of synch with "main".
> 
> Regards,
> Jules

Fantastic, thank you.
I'm rebasing now, and fixing the BigDecimal parsing issue

We can always spilt it into request-a-loan and accept-a-loan and not merge the latter if we're not happy with it.

Z

ZacTredger committed:
Standardise Insufficient Balance exception handling

Two different exceptions were being used for the same job. Also, I moved the
logic for printing the message into AccountBalanceInsufficientException.

ZacTredger committed:
Use 'chain-of-responsibility' pattern in Commands

This de-spaghettifies some commands, like RegisterCommand. And generally enables
code re-use and de-duplication.

I made every Command class have to implement `getSyntax` and added usage/help
messages for each command, implementing issue
`78-add-a-usage-help-for-each-command`.

I wasn't thrilled about the idea of using middleware that mutated the state of
the calling object, but seeing no better alternative, I used this approach to
set fields like `amount`, allowing middleware lower in the stack to rely on
them. Any suggestions as to how to avoid this are most welcome!

michowski commented on docs/protocol.md:
```diff
@@ -381,3 +381,79 @@
+<table>
+  <tr>
+    <td><strong>Description</strong></td>
+    <td colspan="2">Request a peer-to-peer loan. Requests can be seen by all other users.</td>
```
@ZacTredger "Requests can be seen by all other users." why is it so?

michowski replied:
Ah, it's coming from the requirements about P2P. I see now, sorry.

ZacTredger replied:
Haha it's a bit misleading actually, because at the moment they can't actually be _seen_ by _any_ users... because I haven't implemented that.

michowski commented on src/main/java/newbank/server/NewBankClientHandler.java:
```diff
@@ -81,18 +53,8 @@ private boolean processRequest(final String request) {
-    } catch (CommandInvalidSyntaxException e) {
-      handleCommandInvalidSyntax(command);
-    }
+    out.println(getCommand(commandName, tokens).execute());
```
Are you sure this doesn't break the help functionality? Shall we add a few assertions for it in the integration tests?

ZacTredger replied:
Yep, kinda scary that there are not tests for this. I fiddled around with adding parameterized tests using reflection to get the syntax for each command, but I'm too tired to figure it out. :sweat_smile: 

But yes, I'm sure this doesn't break the help functionality! :D

michowski commented on src/main/java/newbank/server/commands/Command.java:
```diff
@@ -1,27 +1,94 @@
+        .findFirst();
+  }
+
+  protected Optional<String> requestingHelp() {
```
What is the value of optional Strings? I believe that for the String data structure we simply concatenate them to obtain the very same effect?

michowski replied:
Or, more specifically, in this case: we can just operate on an empty string?

michowski replied:
I'm very familiar with the concept of `Maybe`/`Optional` in other languages. However, here it seems to be equally dangerous (if not more?) as a standard type that implies the possibility of a `null`?

https://softwareengineering.stackexchange.com/a/309137

> Unfortunately, most languages that now offer Optional haven't abolished null, so you can only profit from the concept by instituting a strict policy of "absolutely no null, ever". Therefore, Optional in e.g. Java is not as compelling as it should ideally be.

michowski commented on src/main/java/newbank/server/commands/Command.java:
```diff
@@ -1,27 +1,94 @@
+   * @return `Optional`ly either a string containing a failure message, or null.
+   */
+  protected Optional<String> applyResponsibilityChain() {
+    return responsibilityChain.stream()
```
Related to the comment below.

So, for example, here: wouldn't this method be much simpler with a simple if check and a string that is null/empty?

michowski commented on src/main/java/newbank/server/commands/Command.java:
```diff
@@ -1,27 +1,94 @@
+   * Apply each of the functions in the responsibility chain. If any of them
+   * return a String (containing a failure message), stop processing.
+   *
+   * @return `Optional`ly either a string containing a failure message, or null.
```
typo

michowski commented on src/main/java/newbank/server/commands/Command.java:
```diff
@@ -1,27 +1,94 @@
+
+  protected Optional<String> incorrectUsage() {
+    String[] expectedTokens = getSyntax().split("\\s");
+    if (tokens.length > expectedTokens.length || incorrectOptionalLiterals(expectedTokens))
```
What happens if a command handles exactly 3 tokens, but only 2 are passed by the user? Is it handled on the `command.execute` level?

ZacTredger replied:
Hmm, this is handled by `incorrectOptionalLiterals`. I'll add a comment to make it clearer how it does this. It's a bit arcane.

michowski commented on src/main/java/newbank/server/commands/Command.java:
```diff
@@ -1,27 +1,94 @@
+          return true;
+      }
+
+      else return tokenCount <= i;
```
Why this `else` formatting? Seems super unusual.

michowski commented on src/main/java/newbank/server/commands/Command.java:
```diff
@@ -1,27 +1,94 @@
+   * returns false.
+   */
+  private boolean incorrectOptionalLiterals(String[] expectedTokens) {
+    Pattern optionalLiteralRegex = Pattern.compile("^\\[.+]$");
```
Would be great to provide an example, as I'm still not fully sure what we are achieving here.

ZacTredger replied:
Ah, this was needed because I'm obviously getting the expected number of tokens from `getSyntax()`, but `ADDACCOUNT <account> [DEFAULT]` introduced the concept of an optional token, so if the command has an 'optional-literal' (e.g./i.e. `"DEFAULT"`), it checks that the  optional token is the correct word.

I'll add the example! :rofl:

michowski replied:
I see, thank you for this great explanation :bow:

michowski reviewed (dismissed):
Hi Zac!

Again, so many great things. Especially so impressive with the tests - thanks a lot. I have left a few comments with small concerns (which might be the result of my poor understanding of Java). However, none of them are blockers.

michowski commented:
Oh, in the meantime - we've got a few conflicts to resolve :))

ZacTredger commented:
@michowski Thank you, this is great.
I think the argument about Optional is really compelling. I'll have a go at refactoring using just Strings instead (with an empty string representing success).
Just managed to resolve all the conflicts locally, but I'll address your comments before pushing again to avoid more middle-of-the-night github notifications.

michowski commented:
@ZacTredger Regarding the optionals - I believe we can merge it as it is currently and then consider a "tech task" to see if strings/nulls are a viable alternative. I don't want to now introduce a blocker for this enormous effort, especially that my suggestion provides no business value. And the tests will be ultra useful in the potential refactoring later on.

ZacTredger committed:
70-request-a-micro-loan

This commit includes tests and an update to the protocol doc, but for the
avoidance of doubt, these features are not implemented here:
- Social responsibility checks for loan-requests
- Any kind of story/justification to accompany the request
- Accepting requests
- Offering loans

I've used a slightly puzzling pattern in the LoanProposals implementation. I
wanted an approach that would be thread-safe for multiple users (relevant once
show-loans/requests and accept-request are implemented), and that would resemble
a DB query so that fewer changes would be needed once persistence is introduced.

This includes a fix for a bug in
91-bug-movemoney-check-and-enforce-minimum-transfer-amount, which allowed users
to deposit/move/pay fractions of pennies.

ZacTredger committed:
73-accept-a-microloan-request

When a request is granted, it is no longer current, and a new microloan is
created.

I've chosen `GRANTLOAN` as the command, over `ACCEPTLOANREQUEST`, which I think
is too verbose (and hard to read!). And to identify the request, each will have
an ID that is unique among active loan-requests, but which, in practice, will
just be the username of the requester (because at the moment users can only have
one active request at a time (and usernames are unique)).

For the avoidance of doubt, the following features have NOT been implemented:
- A `show-loan-requests` command (meaning that for a user to accept a request
they have to know about its existence through other means!)
- A `show-microloans` command (meaning that users aren't able to confirm that
the granted request has become a microloan!)
- Repayments of microloans. This will need some kind of integration with the
`pay-someone` functionality.
- Accruals of microloans. Obviously this would need to be a job that runs
periodically.

juliano60 commented:
@ZacTredger , @michowski 
Had a quick go on the command line and the code gets in a strange state not long after a "QUIT HELP" is issued. It is fairly easy to reproduce. To reproduce do (see screenshot below):
QUIT HELP 
then DEFAULT HELP

Kindest Regards,
Jules

juliano60 commented:
![image](https://user-images.githubusercontent.com/8995677/114823231-20d1d600-9dbb-11eb-9d74-f722e31cc436.png)

juliano60 committed:
66 - Fix QUIT HELP bug

juliano60 commented:
Updated this patch with a fix for that.

Kindest Regards,
Jules

juliano60 commented:
@ZacTredger The applyResponsibility could have returned an Optional instead of an empty String.
Good work. Using a chain of filters is the right approach and a flexible design.
I have not looked too deeply into the "micro-loan" feature per se due to time constraints, I have looked at the other commands.
We have a good battery of tests, let's check in and leave it at that.

Kindest Regards,
Jules

ZacTredger commented:
@michowski & @juliano60 
Thank you both for the awesome reviews!
And [nice catch](https://github.com/Group14-Newbank/newbank/pull/43#issuecomment-820147249) Jules. :grin: 
:tada: :partying_face: 
Z

michowski commented:
Late to the party: Awesome Zac, great job!!! :tada: :rocket:
